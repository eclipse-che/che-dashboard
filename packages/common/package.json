{
  "name": "@eclipse-che/common",
  "version": "7.34.0-SNAPSHOT",
  "repository": "https://github.com/eclipse-che/che-dashboard",
  "license": "EPL-2.0",
  "private": true,
  "main": "lib/index.js",
  "engines": {
    "yarn": ">=1.20.0",
    "node": ">=12"
  },
  "scripts": {
    "build": "yarn clean && tsc -b",
    "watch": "yarn build && tsc -w",
    "clean": "rimraf lib tsconfig.tsbuildinfo"
  },
  "devDependencies": {
    "axios": "^0.21.1",
    "rimraf": "^3.0.2",
    "typescript": "~4.4.2"
  }
}
