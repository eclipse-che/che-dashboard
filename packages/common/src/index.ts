/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import helpers from './helpers';
import * as api from './dto/api';

export * from './dto/cluster-info';
export * from './dto/cluster-config';

export { helpers, api };

export const FACTORY_LINK_ATTR = 'factoryLink';
export const ERROR_CODE_ATTR = 'error_code';
export const KUBECONFIG_MOUNT_PATH = '/tmp/.kube';

const common = {
  helpers,
  api,
};
export default common;
