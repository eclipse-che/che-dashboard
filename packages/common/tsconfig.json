{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src",
    "composite": true
  },
  "include": [
    "src",
  ],
  "exclude": [
    "node_modules",
    "lib",
  ]
}
