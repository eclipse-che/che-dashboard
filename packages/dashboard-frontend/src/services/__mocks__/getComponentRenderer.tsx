/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
