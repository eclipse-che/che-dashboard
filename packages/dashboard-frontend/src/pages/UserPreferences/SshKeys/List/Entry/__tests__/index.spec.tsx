/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { api } from '@eclipse-che/common';
import { TooltipProps } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';

import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

import { Props, SshKeysListEntry } from '..';

jest.mock('@/services/helpers/dates', () => {
  return {
    getFormattedDate: () => 'Aug 27, 4:17 p.m.',
  };
});

// mock the React.Tooltip component
jest.mock('@patternfly/react-core', () => {
  return {
    ...jest.requireActual('@patternfly/react-core'),
    Tooltip: (obj: TooltipProps) => {
      return (
        <div>
          <span>{obj.content}</span>
          {obj.children}
        </div>
      );
    },
  };
});

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

describe('SshKeysListEntry', () => {
  const sshKey: api.SshKey = {
    name: 'sshKey1',
    keyPub: btoa('ssh-public-key'),
    creationTimestamp: new Date('2020-08-27T13:17:34.000Z'),
  };

  const props: Props = {
    sshKey,
    onDeleteSshKey: jest.fn(),
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  test('snapshot', () => {
    const snapshot = createSnapshot(props);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should display the SSH key name', () => {
    renderComponent(props);
    expect(screen.queryByTestId('title')).toHaveTextContent(sshKey.name);
  });

  it('should display the SSH key creation date', () => {
    renderComponent(props);
    expect(screen.queryByTestId('added-on')).toHaveTextContent('Added: Aug 27, 4:17 p.m.');
  });

  describe('dropdown', () => {
    it('should display the dropdown', () => {
      renderComponent(props);
      expect(screen.queryByRole('button', { name: 'Actions' })).not.toBeNull();
    });

    it('should reveal the Delete button', async () => {
      renderComponent(props);
      const dropdown = screen.getByRole('button', { name: 'Actions' });
      await userEvent.click(dropdown);

      const deleteItem = screen.queryByRole('menuitem', { name: 'Delete' });
      expect(deleteItem).not.toBeNull();

      await userEvent.click(deleteItem!);
      expect(props.onDeleteSshKey).toHaveBeenCalledWith(sshKey);
    });
  });

  describe('copy to clipboard', () => {
    it('should display the Copy to Clipboard button', () => {
      renderComponent(props);
      expect(screen.queryByTestId('copy-to-clipboard')).not.toBeNull();
    });

    it('should copy the SSH key to the clipboard', async () => {
      window.prompt = jest.fn();
      jest.useFakeTimers();
      const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

      renderComponent(props);
      const copyToClipboard = screen.getByTestId('copy-to-clipboard');

      // initial tooltip state
      expect(screen.queryByText('Copied!')).toBeNull();
      expect(screen.queryByText('Copy to clipboard')).not.toBeNull();

      await user.click(copyToClipboard);

      // 500ms after click
      await jest.advanceTimersByTimeAsync(500);
      expect(screen.queryByText('Copied!')).not.toBeNull();
      expect(screen.queryByText('Copy to clipboard')).toBeNull();

      await user.click(copyToClipboard);

      // 1000ms after after the second click
      await jest.advanceTimersByTimeAsync(500);
      expect(screen.queryByText('Copied!')).not.toBeNull();
      expect(screen.queryByText('Copy to clipboard')).toBeNull();

      // 5000 after the last click
      await jest.advanceTimersByTimeAsync(5000);
      expect(screen.queryByText('Copied!')).toBeNull();
      expect(screen.queryByText('Copy to clipboard')).not.toBeNull();

      await jest.runOnlyPendingTimersAsync();
      jest.useRealTimers();
    });
  });
});

function getComponent(props: Props) {
  return <SshKeysListEntry {...props} />;
}
