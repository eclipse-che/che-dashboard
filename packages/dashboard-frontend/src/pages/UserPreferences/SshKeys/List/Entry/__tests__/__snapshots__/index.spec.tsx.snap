// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SshKeysListEntry snapshot 1`] = `
<article
  className="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-1"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe={true}
  id=""
>
  <div
    className="pf-c-card__header"
  >
    <div
      className="pf-c-card__title"
      data-testid="title"
    >
      sshKey1
    </div>
    <div
      className="pf-c-card__actions"
    >
      <div>
        <span>
          Copy to clipboard
        </span>
        <button
          aria-disabled={false}
          aria-label={null}
          className="pf-c-button pf-m-link"
          data-ouia-component-id="OUIA-Generated-Button-link-1"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          data-testid="copy-to-clipboard"
          disabled={false}
          name="Copy to Clipboard"
          onClick={[Function]}
          role={null}
          type="button"
        >
          <span
            className="pf-c-button__icon pf-m-start"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="pf-c-dropdown pf-m-align-right"
        data-ouia-component-id="OUIA-Generated-Dropdown-1"
        data-ouia-component-type="PF4/Dropdown"
        data-ouia-safe={true}
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="Actions"
          className="pf-c-dropdown__toggle pf-m-plain"
          disabled={false}
          id="pf-dropdown-toggle-id-0"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 192 512"
            width="1em"
          >
            <path
              d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    className="pf-c-card__footer"
  >
    <div
      className="pf-c-content"
    >
      <small
        className=""
        data-ouia-component-id="OUIA-Generated-Text-1"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe={true}
        data-pf-content={true}
        data-testid="added-on"
      >
        Added: 
        Aug 27, 4:17 p.m.
      </small>
    </div>
  </div>
</article>
`;
